File: MonitoraUFF-feat-frontend-tickets/backend/core/map/region.py
from abc import ABC, abstractmethod
from typing import Tuple, Dict, Any, Optional, TypedDict
from enum import Enum

Extend = Tuple[float, float, float, float]
"""A collection of four values representing an array of 4 numbers that define the geographic extent of the map in the EPSG:4326 projection.
x = longitude, y = latitude [minX, minY, maxX, maxY]"""

Polygon = Dict[str, Any]
"""
Data structure that contains information about the map polygon.

Example:
    {
        "type": "Feature",
        "geometry": {
            "type": "LineString",
            "coordinates": []
        }
    }

Attributes:
    - type (str): A string indicating the type of the polygon. Common values include 'Feature'.
    - geometry (dict): A dictionary that describes the polygon's geometry, which includes type and coordinates.
"""


class RegionType(Enum):
    """
    Enumeration that represents the types of regions for a region creation operation.

    The `RegionType` enumeration defines two possible values: `EXTERNAL` and `INTERNAL`, which indicate
    whether a region is external or internal, respectively.
    """
    EXTERNAL = "EXTERNO"
    INTERNAL = "INTERNO"

class LevelHierarchyType(Enum):
    """
    Enumeration that represents the types of zooms for a region creation operation.

    The `LevelHierarchyType` enumeration defines five possible values: `COUNTRY`,`STATE`,`CITY`,`CAMPUS` and `BUILD`, which indicate the zoom of the region.
    """
    COUNTRY = "PAIS"
    STATE = "ESTADO"
    CITY = "CIDADE"
    CAMPUS = "CAMPUS"
    BUILD = "PREDIO"

class NewRegion(TypedDict):
    """ Dict that represents the fields returned in a region creation operation """
    id: int
    success: bool

class UpdatedRegion(TypedDict):
    """ Dict that represents the fields returned in a region update operation """
    id: int
    success: bool

class DeletedRegion(TypedDict):
    """ Dict that represents the fields returned in a region delete operation """
    id: int
    success: bool

class RegionManagerAbstract(ABC):

    @abstractmethod
    def create_region(self, *args, name: str, acronym: str, description: str, extend: Extend, type_region: RegionType, image: str, polygon: Polygon,level_type:LevelHierarchyType,upper_region: Optional[int], **kwargs: Any) -> NewRegion:
        raise NotImplementedError

    @abstractmethod
    def get_root_regions(self, *args, **kwargs: Any) -> Tuple[Any, Any]:
        raise NotImplementedError

    @abstractmethod
    def get_all_regions(self, *args, **kwargs: Any) -> Tuple[Any, Any]:
        raise NotImplementedError

    @abstractmethod
    def get_root_region_by_id(self, *args, id: str, **kwargs: Any) ->Tuple[Any, Any]:
        raise NotImplementedError

    @abstractmethod
    def get_region_by_id(self, *args, id: str, **kwargs: Any) ->Tuple[Any, Any]:
        raise NotImplementedError

    @abstractmethod
    def delete_region_by_id(self, *args, id: str, **kwargs: Any) -> DeletedRegion:
        raise NotImplementedError

    @abstractmethod
    def update_region_by_id(self, *args, id: str, **kwargs: Any) -> UpdatedRegion:
        raise

File: MonitoraUFF-feat-frontend-tickets/frontend/web/src/models/region.ts
import { createRegionZoom } from "./zoom";

export interface Region {
  id: number;
  name: string;
  acronym: string;
  image: string | null;
  description: string;
  level_type: keyof typeof createRegionZoom;
  type_region: "EXTERNO" | "INTERNO";
  polygon: PolygonGeoJson;
  upper_region: number;
  extend: string[];
}

export interface PolygonGeoJson {
  type: "FeatureCollection";
  features: PolygonFeature[];
}

interface PolygonFeature {
  type: "Feature";
  geometry: PolygonGeometry;
  properties: {
    id: number;
    name: string;
    type: string;
  };
}

interface PolygonGeometry {
  type:"LineString" | "Point";
  coordinates: number[];//NUMBER[][][]
}

export interface RegionHierarchy
  extends Pick<Region, "id" | "name" | "upper_region"> {
  children?: RegionHierarchy[];
}

export interface RegionAdd {
  name: string;
  acronym: string;
  description: string;
  image: string | null;
  extend: string[];
  type_region: string;
  level_type: string;
  polygon: object;
  upper_region: string;
}

export const REGION_DATA_INITIAL_VALUE: Region = {
  acronym: "",
  description: "",
  extend: [""],
  id: 0,
  image: null,
  level_type: "CAMPUS",
  name: "",
  polygon: {
    features: [
      {
        geometry: {
          coordinates: [[[0, 0, 0]]],
          type: "Polygon",
        },
        properties: {
          id: 0,
          name: "",
          type: "",
        },
        type: "Feature",
      },
    ],
    type: "FeatureCollection",
  },
  type_region: "EXTERNO",
  upper_region: 0,
};